### DISCUSSION TITLE
How to use spacy train to add entities to an existing custom NER model? (Spacy v3.0)

### DISCUSSION THREAD

**Discussion by chaitanyamalik** [explosion/spaCy#8473] [Help: Coding & Implementations] [answered]
So, I am currently implementing a custom NER model interface where a user can interact with a frontend application to add custom entities to train a spacy model.

I want to use spacy train (CLI) to take an existing model (custom NER model) and add the keyword and entity specified by the user, to that model. (Instead of training the whole model again). Please help as I can't find this anywhere in the documentation.

For example, let's say I have a model that is already trained for a custom entity of FOOD. (Pizza, Pasta, Bread, etc…). Now I want to take this existing model, and train it for a new entity called DRINKS with keywords like Coca-Cola, Pepsi, Juice, etc… Using spacy train command for spacy v3.0.

The spacy train command that I am using currently is as follows:

`python -m spacy train config.cfg --output ./output --paths.train ./train.spacy --paths.dev ./train.spacy`

I load the model for prediction using:

` nlp1 = spacy.load(R".\output\model-best")`

As of now, I was training the model for new entities manually. Below is the code to find keywords in my training data and output a JSON format for training data (old format).

`import re

keyword = ["outages","updates","negative star","worst"]
entity = ["PROBLEM","PROBLEM","COMPLAINT","COMPLAINT"]

train = []

for text in df.text:

    for n in range(0,len(keyword)):
    
        start_index = []
        end_index = []

        start_index = [m.start() for m in re.finditer(keyword[n], str(text))]

        if(start_index):

            end_index = [m+len(keyword[n]) for m in start_index]

            for i in range(0,len(start_index)):

                train.append((text,{"entities": [(start_index[i],end_index[i],entity[n])]}))

After this, I converted my json format into .spacy format with the below code.

```from tqdm import tqdm
from spacy.tokens import DocBin

db = DocBin()

for text, annot in tqdm(train): 
    doc = nlp.make_doc(text) 
    ents = []
    for start, end, label in annot["entities"]: 
        span = doc.char_span(start, end, label=label, alignment_mode="contract")
        if span is None:
            print("Skipping entity")
        else:
            ents.append(span)
    doc.ents = ents
    db.add(doc)

db.to_disk("./train.spacy")```

**Comment by polm** [comment, marked as answer]
Here's [your question on Stack Overflow](https://stackoverflow.com/questions/68083466/how-to-use-spacy-train-to-add-entities-to-an-existing-custom-ner-model-spacy-v). I also answered there, but briefly, you want to do **online learning** but that's not supported in spaCy or neural frameworks in general. 

You can use a rule-based method or train small models as the user gives you more data. 

Writing a custom training loop will not allow you to work around this. If you update the model just with new data it will forget old info and you'll get weird results.

**Reply by chaitanyamalik** [reply]
Hi @polm,

Thanks a lot for your help!!
After your detailed answer to my Stackoverflow question, I've replied with a certain query over there.
[Stackoverflow link](https://stackoverflow.com/questions/68083466/how-to-use-spacy-train-to-add-entities-to-an-existing-custom-ner-model-spacy-v)
